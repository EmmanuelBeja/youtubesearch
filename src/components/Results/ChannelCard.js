import React from 'react'

import { Card, CardText, CardBody, CardTitle, CardImg, Row, Col } from 'reactstrap'

const ChannelCard = ({ data, handleRedirectToChannel }) => {
  return (
    <Card
      className="channel-card"
      onMouseDown={() => handleRedirectToChannel(data.snippet.channelId)}
    >
      <CardBody>
        <Row>
          <Col md="4" sm="12" xs="12" className="channel-card__thumbnail-container">
            <CardImg
              className="channel-card__thumbnail"
              top
              width="100%"
              src={data.snippet.thumbnails.default.url}
              alt="channel thumbnail"
            />
          </Col>
          <Col md="8" sm="12" xs="12" className="">
            <CardTitle tag="h5" className="text-truncate channel-name">
              {data.snippet.channelTitle}
            </CardTitle>
            <CardText>{data.snippet.description}</CardText>
          </Col>
        </Row>
      </CardBody>
    </Card>
  )
}

export default ChannelCard
