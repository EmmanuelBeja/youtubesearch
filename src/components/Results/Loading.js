import React from 'react'

const Loading = () => {
  return (
    <div className="text-center loading-container">
      <img src="/sync.svg" className="loading-img" alt="loading" />
      <br />
      <br />
      <p>Loading...</p>
    </div>
  )
}

export default Loading
