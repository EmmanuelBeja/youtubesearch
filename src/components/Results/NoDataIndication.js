import React from 'react'

const NoDataIndication = () => {
  return (
    <div className="text-center">
      <img
        src={`idk${Math.floor(Math.random() * 4 + 1)}.gif`}
        className="data-empty"
        alt="data empty"
      />
      <br />
      <br />
      <h3>
        <b>No results found</b>
      </h3>
      <p>Try different keywords or remove search filters</p>
    </div>
  )
}

export default NoDataIndication
