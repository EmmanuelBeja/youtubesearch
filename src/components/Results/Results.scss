.card {
  border: none;
  cursor: pointer;

  &:hover {
    background: whitesmoke;
  }

  &-subtitle {
    color: grey;
  }

  .channel-name {
    color: #2196f3;

    &:hover {
      color: #2196f3;
    }
  }
}

.channel-card {
  border-bottom: 1px solid whitesmoke;
  margin-bottom: 10px;

  &__thumbnail {
    height: 100px;
    width: 100px;
    border-radius: 50%;

    &-container {
      display: flex;
      justify-content: center;
      align-items: center;
    }
  }
}

.data-empty,
.loading-img {
  height: 170px;
  margin-top: 50px;
}
