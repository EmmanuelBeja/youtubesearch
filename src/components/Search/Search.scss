.navbar {
  border-top: 5px solid #fe0101;
  border-bottom: 1px solid whitesmoke;
  background-color: white !important;

  .search-section {
    display: flex;
    justify-content: center;
    padding: 0px 50px;
    width: 100%;

    .input-group {
      width: 30%;
      transition: all 0.4s;
      border: 1px solid #ced4da;
      border-radius: 5px;

      &:focus-within {
        width: 35%;

        .autocomplete {
          display: block;
          transition: all 9s;
        }
      }

      @media only screen and (max-width: 576px) {
        width: 98%;

        &:focus-within {
          width: 100%;
        }
      }

      &-prepend {
        background-color: #3f51b514;
        border-radius: 5px 0px 0px 5px !important;
        display: flex;
        justify-content: center;
        align-items: center;
        width: 30px;
      }

      .search__autocomplete > .autocomplete {
        position: absolute;
        left: 0;
        top: 50px;
        right: 0;
        border-radius: 5px;
      }

      .autocomplete {
        display: none;
        background-color: white;
        color: #444;
        box-shadow: 0 1px 2px rgba(0, 0, 0, 0.15);
        font-size: 0.875rem;
        text-align: left;

        &__info {
          color: grey;
          padding: 1rem 2rem 0 2rem;
          font-size: 0.5rem;
        }

        &__item {
          padding: 0.5rem 2rem;
          cursor: pointer;

          &:hover {
            background-color: #eee;
          }
        }
      }

      .form-control {
        border: transparent;
        border-radius: 0px 5px 5px 0px !important;
        background-color: #3f51b514;
        margin-left: 0px !important;

        &:focus {
          outline: none;
          box-shadow: none;
          border-color: transparent;
        }
      }

      &-append {
        button {
          border: none;

          &:hover {
            background-color: white;
            color: black;
          }
        }
      }
    }
  }
}
