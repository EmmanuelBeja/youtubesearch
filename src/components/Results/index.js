import React from 'react'
import { useSelector } from 'react-redux'

import ChannelCard from './ChannelCard'
import VideoCard from './VideoCard'
import NoDataIndication from './NoDataIndication'
import Loading from './Loading'

import './Results.scss'

const Results = () => {
  const { loading, results } = useSelector((state) => state.search)

  // if (loading) {
  //   return <Loading />
  // }

  const handleRedirectToChannel = (channelId) =>
    window.open(`https://www.youtube.com/channel/${channelId}`)

  return (
    <div className="results-container">
      {loading ? (
        <Loading />
      ) : (
        <>
          {results.map((data, index) => (
            <div key={index}>
              {data.id.kind === 'youtube#channel' ? (
                <ChannelCard data={data} handleRedirectToChannel={handleRedirectToChannel} />
              ) : (
                <VideoCard data={data} handleRedirectToChannel={handleRedirectToChannel} />
              )}
            </div>
          ))}
          {!results || (results.length === 0 && <NoDataIndication />)}
        </>
      )}
    </div>
  )
}

export default Results
