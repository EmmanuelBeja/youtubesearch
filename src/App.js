import React from 'react'
import { Provider } from 'react-redux'
import store from './store'

import { ToastContainer } from 'react-toastify'
import 'react-toastify/dist/ReactToastify.min.css'
import { Container } from 'reactstrap'
import Search from './components/Search'
import Results from './components/Results'

import './App.scss'

const App = () => {
  return (
    <Provider store={store}>
      <Container className="app-container">
        <Search />
        <Results />
      </Container>
      <ToastContainer />
    </Provider>
  )
}

export default App
