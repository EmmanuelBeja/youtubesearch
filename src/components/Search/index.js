import React, { useState, useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'

import { Navbar, InputGroup, InputGroupAddon, Input } from 'reactstrap'
import { useDebounce } from 'use-debounce'
import { faSearch, faCircleNotch } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { toast } from 'react-toastify'
import RecentSearches from './RecentSearches'

import { searchActions } from './actions'

import './Search.scss'

const Search = () => {
  const dispatch = useDispatch()
  const { loading } = useSelector((state) => state.search)

  const [searchInput, setSearchInput] = useState('')
  const [debouncingText, setDebouncingText] = useState('')
  const [debouncedText] = useDebounce(debouncingText, 2000)
  const [inputError, setInputError] = useState(false)
  const [searches, setSearches] = useState([])

  useEffect(() => {
    debouncedText && handleSearch(debouncedText)
  }, [debouncedText])

  const handleSearch = (input) => {
    if (!input) {
      setInputError(true)
      toast.error('Please type something to search')
      return
    }
    // add searched item to recent searches
    const updatedSearches = searches
    updatedSearches.push(input)

    setSearches(updatedSearches)
    dispatch(searchActions.getSearchResults(input))
    setInputError(false)
  }
  // {{ size: 6, offset: 3 }}
  return (
    <Navbar color="light" light expand="md" className="fixed-top">
      <div className="search-section">
        <InputGroup>
          <InputGroupAddon addonType="prepend">
            {loading ? (
              <FontAwesomeIcon icon={faCircleNotch} pulse />
            ) : (
              <FontAwesomeIcon icon={faSearch} />
            )}
          </InputGroupAddon>
          <Input
            className="form-control"
            name="searchInput"
            placeholder="Type here..."
            value={searchInput}
            onChange={(e) => {
              setDebouncingText(e.target.value)
              setSearchInput(e.target.value)
            }}
            invalid={inputError}
            autoComplete="off"
          />
          <RecentSearches searches={searches} handleClick={handleSearch} />
        </InputGroup>
      </div>
    </Navbar>
  )
}

export default Search

// <Row>
//   <Col
//     lg="12"
//     md="12"
//     sm="12"
//     xs="12"
//     className="search-section shadow"
//   >
//     <InputGroup>
//       <InputGroupAddon addonType="prepend">
//         {loading ? (
//           <FontAwesomeIcon icon={faCircleNotch} pulse />
//         ) : (
//           <FontAwesomeIcon icon={faSearch} />
//         )}
//       </InputGroupAddon>
//       <Input
//         className="form-control"
//         name="searchInput"
//         placeholder="Type here..."
//         value={searchInput}
//         onChange={(e) => {
//           setDebouncingText(e.target.value)
//           setSearchInput(e.target.value)
//         }}
//         invalid={inputError}
//         autoComplete="off"
//       />
//
//       <RecentSearches searches={searches} handleClick={handleSearch} />
//     </InputGroup>
//   </Col>
// </Row>
