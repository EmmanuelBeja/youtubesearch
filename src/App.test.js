import React from 'react'
import { shallow, mount } from 'enzyme'
import App from './App'
import Search from './components/Search'
import Results from './components/Results'
import { ToastContainer } from 'react-toastify'

describe('App', () => {
  const shallowWrapper = shallow(<App />)
  const mountWrapper = mount(<App />)

  it('should render correctly', () => {
    expect(shallowWrapper).toMatchSnapshot()
    expect(shallowWrapper.find('.app-container').length).toEqual(1)
  })
  it('should contain Search component', () => {
    expect(mountWrapper.find(<Search />)).toBeTruthy()
  })
  it('should contain Results component', () => {
    expect(mountWrapper.contains(<Results />)).toBeTruthy()
  })
  it('should contain ToastContainer component', () => {
    expect(mountWrapper.find(<ToastContainer />)).toBeTruthy()
  })
})
