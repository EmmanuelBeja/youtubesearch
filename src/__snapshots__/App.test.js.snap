// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Container
    className="app-container"
    tag="div"
  >
    <Search />
    <Results />
  </Container>
  <ToastContainer
    autoClose={5000}
    closeButton={[Function]}
    closeOnClick={true}
    draggable={true}
    draggableDirection="x"
    draggablePercent={80}
    hideProgressBar={false}
    newestOnTop={false}
    pauseOnFocusLoss={true}
    pauseOnHover={true}
    position="top-right"
    role="alert"
    rtl={false}
    transition={[Function]}
  />
</Provider>
`;
